### PR TITLE
Make ``GeoHash`` coordinate conversions ~40-100x faster

### PR DESCRIPTION
**Pull request by PaulusParssinen**
> [!NOTE]
> Credits for the quantization approach to https://mmcloughlin.com/posts/geohash-assembly (by @mmcloughlin)
> and de-quantization method to https://github.com/georust/geohash, big thank you for everyone contributing to both efforts! 

This PR originally was meant to be simply tiny optimization to avoid [unnecessary heap allocations on the coordinate conversions](https://github.com/microsoft/garnet/pull/348/commits/6a6a6ef2c1d808a93f22d467303664b86857c1c1) but I had change of plans when I discovered an article about much faster (de)[quantization method](https://mmcloughlin.com/posts/geohash-assembly) by Michael McLoughlin and this ended up being `GeoHash` class rewrite. Please see his great article and explanation.

Much of the commits after https://github.com/microsoft/garnet/pull/348/commits/564891a01678cd8b7b6a4772e10f37d838e076ba is me discovering new exciting details about IEEE-754 floating-point rounding, confusing reference implementations (it turns out, there's no single correct way to reference a location on earth 😅) and other fun Geohash corner-cases. 

However, more notably there's __tiny__ error in the quantization approach described in the McLoughing's excellent article which [I describe in this comment](https://github.com/microsoft/garnet/pull/348#issuecomment-2096096777) and fix in [`f364a96c..b4ec75c`](https://github.com/microsoft/garnet/pull/348/files/f364a96ca1fdbf0aaf69c7c979d3daa01d9826b6..b4ec75cf1214574f2a05369b033c10a2f59d5588). Gist of it is that in IEEE-754 double precision format, the exponent assumption of `1023` does not hold for `2.0`, which the maximum coordinate values get rounded to.

There _should_ be __no other__ behavioral differences than the last character of textual representation, which one of the existing tests _already_ ignored understandably due to nuances around the precision of it. The textual representation ~~is~~ should simply be base-32 encoded (custom geohash alphabet) of the integer that represents a specific GeoHash where more characters means more bits of precision. Every encoded character gives 5-bits of precision. Very simple right! <sub>I wish it was</sub>

Garnet (and Redis) chose 52-bit precision. Why? Because we want to store the geohash integer in a sorted set, which allows us to utilize the geohash structure and do really efficient range queries. Now because we want to reuse the existing sorted set we need to safely store the geohash integer in IEEE-754 double precision score which we can do that safely when our integer is in range `[(-2)^53. 2^53]`. 

Now, we can still encode this 52-bit precision integer in the base-32 alphabet just fine. The remaining 55-52=3 bits will be zero and that's what Garnet has done, but Redis _seems_ to zero out the last 2 bits and output the last character of the textual representation as `0`.

Redis made another choice too, which is that they chose to follow EPSG:3875 where:
* Valid longitudes are from -180 to 180 degrees.
* Valid latitudes are from -85.05112878 to 85.05112878 degrees.

This makes sense as the poles cause all sorts of problems in GIS applications.

<details>
<summary><a href="https://wiki.openstreetmap.org/wiki/Web_Mercator">Confusing history behind Web Mercator</a></summary>

image

</details>

These limits however do not apply to "standard" geohash format which valid latitudes are `[-90, 90]`. This difference also adds nuance to geohash implementations. Garnet _currently_ accepts coordinates per the geohash limitations.

This PR also introduces `AsciiUtils` in `Garnet.common` as place to share common ASCII manipulation logic (acts more like polyfill for `Ascii` class in .NET 6 and a place where we can optimize further than what BCL's currently does).

This PR is strictly focused on `GeoHash` class alone. I'll remove more unnecessary allocations in the __actual__  RESP protocol parsing & encoding in follow-up. There's some vectorization opportunities here and the distance math can be definitely improved too.

<details>
<summary><h3>Benchmarks (Tiger Lake, i5-1135G7)</h3></summary>

```
BenchmarkDotNet v0.13.12, Ubuntu 22.04.4 LTS (Jammy Jellyfish)
11th Gen Intel Core i5-1135G7 2.40GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.204
  [Host] : .NET 8.0.4 (8.0.424.16909), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI
  .NET 6 : .NET 6.0.29 (6.0.2924.17105), X64 RyuJIT AVX2
  .NET 8 : .NET 8.0.4 (8.0.424.16909), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI

Server=True  
```

### `main`
| Method                 | Job    | Mean     | Error   | StdDev  | Gen0   | Allocated |
|----------------------- |------- |---------:|--------:|--------:|-------:|----------:|
| GeoToLongValue         | .NET 6 | 278.3 ns | 0.51 ns | 0.43 ns | 0.0191 |      80 B |
| GeoToLongValue         | .NET 8 | 260.0 ns | 0.85 ns | 0.80 ns | 0.0191 |      80 B |
| GetCoordinatesFromLong | .NET 6 | 429.1 ns | 1.30 ns | 1.22 ns | 0.0801 |     336 B |
| GetCoordinatesFromLong | .NET 8 | 213.8 ns | 0.85 ns | 0.75 ns | 0.0801 |     336 B |
| GetGeoHashCode         | .NET 6 | 810.1 ns | 4.79 ns | 4.24 ns | 0.2537 |    1064 B |
| GetGeoHashCode         | .NET 8 | 619.1 ns | 1.88 ns | 1.76 ns | 0.2537 |    1064 B |

### PR
| Method                 | Job    | Mean      | Error     | StdDev    | Gen0   | Allocated |
|----------------------- |------- |----------:|----------:|----------:|-------:|----------:|
| GeoToLongValue         | .NET 6 |  5.556 ns | 0.0156 ns | 0.0146 ns |      - |         - |
| GeoToLongValue         | .NET 8 |  1.888 ns | 0.0156 ns | 0.0130 ns |      - |         - |
| GetCoordinatesFromLong | .NET 6 |  5.074 ns | 0.0038 ns | 0.0031 ns |      - |         - |
| GetCoordinatesFromLong | .NET 8 |  1.215 ns | 0.0038 ns | 0.0034 ns |      - |         - |
| GetGeoHashCode         | .NET 6 | 15.944 ns | 0.2487 ns | 0.2077 ns | 0.0014 |      48 B |
| GetGeoHashCode         | .NET 8 | 15.933 ns | 0.0331 ns | 0.0294 ns | 0.0012 |      48 B |

</details>

<details>
<summary><h3>Benchmarks (Zen 2, Ryzen 7 3700X)</h3></summary>

```
BenchmarkDotNet v0.13.12, Windows 10 (10.0.19045.4170/22H2/2022Update)
AMD Ryzen 7 3700X, 1 CPU, 16 logical and 8 physical cores
.NET SDK 9.0.100-preview.3.24204.13
  [Host] : .NET 8.0.4 (8.0.424.16909), X64 RyuJIT AVX2
  .NET 6 : .NET 6.0.27 (6.0.2724.6912), X64 RyuJIT AVX2
  .NET 8 : .NET 8.0.4 (8.0.424.16909), X64 RyuJIT AVX2

Server=True
```

### `main`
| Method                 | Job    | Mean     | Error   | StdDev  | Gen0   | Allocated |
|----------------------- |------- |---------:|--------:|--------:|-------:|----------:|
| GeoToLongValue         | .NET 6 | 290.1 ns | 3.08 ns | 2.88 ns | 0.0095 |      80 B |
| GeoToLongValue         | .NET 8 | 269.4 ns | 2.14 ns | 1.90 ns | 0.0095 |      80 B |
| GetCoordinatesFromLong | .NET 6 | 510.0 ns | 9.71 ns | 9.54 ns | 0.0401 |     336 B |
| GetCoordinatesFromLong | .NET 8 | 195.3 ns | 3.94 ns | 3.49 ns | 0.0401 |     336 B |
| GetGeoHashCode         | .NET 6 | 877.3 ns | 6.87 ns | 6.42 ns | 0.1268 |    1064 B |
| GetGeoHashCode         | .NET 8 | 525.6 ns | 6.46 ns | 6.04 ns | 0.0982 |     824 B |

### PR
| Method                 | Job    | Mean      | Error     | StdDev    | Gen0   | Allocated |
|----------------------- |------- |----------:|----------:|----------:|-------:|----------:|
| GeoToLongValue         | .NET 6 |  5.646 ns | 0.0306 ns | 0.0255 ns |      - |         - |
| GeoToLongValue         | .NET 8 |  5.829 ns | 0.0537 ns | 0.0502 ns |      - |         - |
| GetCoordinatesFromLong | .NET 6 |  5.507 ns | 0.0583 ns | 0.0545 ns |      - |         - |
| GetCoordinatesFromLong | .NET 8 |  5.928 ns | 0.0423 ns | 0.0396 ns |      - |         - |
| GetGeoHashCode         | .NET 6 | 16.286 ns | 0.0968 ns | 0.0808 ns | 0.0003 |      48 B |
| GetGeoHashCode         | .NET 8 | 15.721 ns | 0.1790 ns | 0.1587 ns | 0.0003 |      48 B |

<details>
<summary>Zen 2 PDEP/PEXT "erratum" in action</summary>

|Method|Job|Mean|Error|StdDev|Allocated|
|--- |--- |--- |--- |--- |--- |
|GeoToLongValue|.NET 6 w/ UsePdepPext|64.98 ns|0.132 ns|0.103 ns|-|
|GeoToLongValue|.NET 8 w/ UsePdepPext|64.95 ns|0.383 ns|0.359 ns|-|
|GetCoordinatesFromLong|.NET 6 w/ UsePdepPext|62.87 ns|0.137 ns|0.128 ns|-|
|GetCoordinatesFromLong|.NET 8 w/ UsePdepPext|62.92 ns|0.090 ns|0.084 ns|-|

</details>

</details>

<sup>See that 48B heap allocation for the final string in ``GetGeoHashCode``, we don't need that :)</sup>

### CODE REVIEW

**Comment by PaulusParssinen** [comment]
wip: explain comment

### CanEncodeAndDecodeCoordinates

| Label                      | Latitude (F16)      | Longitude (F16)      | Lat Error (F16)    | Lon Error (F16)     |
| -------------------------- | ------------------- | -------------------- | ------------------ | ------------------- |
| Test input                 | 30.5388942218       | 104.0555758833       | -                  | -                   |
| Test IEEE-754 repr.        | 30.5388942217999997 | 104.0555758832999942 | -                  | -                   |
| main                       | 30.5388928949832916 | 104.0555772185325623 | 0.0000013268167081 | -0.0000013352325681 |
| PR                         | 30.5388915538787842 | 104.0555745363235474 | 0.0000026679212155 | 0.0000013469764468  |
| GetGeoErrorByPrecision(52) | -                   | -                    | 0.0000026822090149 | 0.0000053644180298  |

**Comment by PaulusParssinen** [comment]
Premature optimization is ~~root of all evil~~ fun!

Few test failures left, need to dive deeper into some rounding corner-cases.

File: libs/server/Objects/SortedSetGeo/GeoHash.cs [hunk 1]
```diff
@@ -11,53 +15,70 @@ namespace Garnet.server
...
-        //The PI/180 constant
-        static readonly double degreesToRadians = 0.017453292519943295769236907684886;
-
-        //The "Geohash alphabet" (32ghs) uses all digits 0-9 and almost all lower case letters except "a", "i", "l" and "o".
-        //This table is used for getting the "standard textual representation" of a pair of lat and long.
-        static readonly char[] base32chars = new char[] { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'j', 'k', 'm', 'n', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z' };
+        /// <summary>
+        /// The number of bits used for the precision of the geohash.
+        /// </summary>
+        public const int BitsOfPrecision = 51;
```
Comment (PaulusParssinen) [inline libs/server/Objects/SortedSetGeo/GeoHash.cs:27]: Oops, was experimenting stuff. Let's see the actual test failures
```suggestion
        public const int BitsOfPrecision = 52;
```

**Comment by PaulusParssinen** [comment]
Added ``USE_PDEP_PEXT`` to allow developers who know their CPU has the non-microcode-emulated version. [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8B6YgAgGEIAHATygEsBzACwzIAowBKMgWQZgoEXBABm7KlGrRsGBhAB2AOgCwAKFJkAMoJiLcMACZkAroqMwoZDCxj8AkgBUyAGz0GYa9RoDElsQZFewBVAGUAUQB9AAUAEQiY2IiADScNDQABACYARj8GMTJw6PjE5LSsgAYyTNzlACVzeXwvB0UMRgNBXGUUgA4kAG4/fSNCjPVFbFbcamwwewBxbChgjGVDKAA3Kw0Abw0yY7JScgAeXFN8fFXaAD4jk7OyCMVICyCmMmwLMktIONFN9WrYIEZcGQxNAyEsYCJ5vJsK4nscXudiFcbndHuoTrUAMy1XJIWrZWHwgASeBYqLIhzx+Oe5CoBk62CCGEhYmE+FeMTCSxAAE4qlVhbkieREoKQASAOz9ACspzIAHlBRE1SAULkxflGUyyNRGNs5PYPrh2EYIKZgK57NolEwGBhTJYBIoyABeMhwXL9KrKKojQ1Mk0MM0YC1KK3/W32x3O13umB8bAIH1kANBkOTI3HCNRmNs+N2h06OQpj1BLNwUXB0MF42m81kS3WhMV7RVt0ejNZht5nxh/HorG3KAPOljrROOxkRTXYBWMjiMjAV2Q0yGEzQ6y2ewmmBgBi4BRe9eHshMeEsGneZvozHXSfT0cnTJEjtkTlkABCW5qmIMSwKe55KFmSrZKGM7MhcE44nBaJzguDrArYa5FNet4QPeuAsGQABEVq/EYqwmNGCBusiZCwMehgdFWShETekb6DYC6XDAFqwGI3pEYBXLAaBJ5nheRH3Ju7DHuBF4YoYMD3L+kLQY+BbPohU64s2X7trG7B/lQljaPoTCYb6uS5LBH4oQhr5IbZpxaG8gIwJC15utQFbrhw5zzFAMx8YuMwwAJri9qmRHEPcaBkP5qxBTARTTK04XJn2MDRfcvCQNAQLmh5EA/GYigMAAjqY9jQXA0m/h0MC3lAyHORcL7YtpLV6XUpKuM6FIQE4EBOsCABqyJVRwNrlvYEXyJlcXTYmbgZam3AtQyzb4oUnAAIQcD282pp68W+nN1b2AAZJdbiRZYp2Vkd/YILwhCEC1zb7SNLqZSd5xnat93XStwIXQ930XemL3rU5+KbVtRqZPKfq5NoTYI2QAC++YYz+S3dnIQ0hGVGANL8t4NGJJoQGAtGWaqOaNh9Jx412SaKETJNk8CMCU6e1O064Wa5KqBJILmNkY85lCwOMXI2MVbAYNQuAgKQNxgH1pisO4KiQPgxAyFauDELh+EsHAeCGPg9q0FCMLXpVvzyAAXsxXrYNQ1MLCwu04wjqExvlQSFT8sBkE7HQMC7xgbnbgRQHGtxe18nEwHyGDFde5iuvV0Y7LRADaVRxbkAC6PwWMzKFp16+CmK48jebQqfXtkAB6BLZMoZAAGIwjACAzN5MBxZnN4wOwXe1bnkeu7HnKNau9GwIxGDu1hae3U9MDV9LADuLCCIRZ71XR5P2IX9Yl2QooV/v9j77argmDaIB76QH/kHNACKpjO9HWOvoxB9WgBwbuVRO7kgAFScHOplMgABqW+uZeDkEZlUGGUtP6w2ZGQSkEAH67CgHFUQrQyD71dIRTWMBiFkFuJhDgYQeL8DAE6bWLBdYAHJIRMFgHIMO8gaE/GABAXY9sDwLksOvBgrhcC7SwRjLQZE44iMyDUa8DgIjaLgPKJUKA6IwAYvode8hILrnxvAWS4lIIgIgFWYEcAZB/iXDbKwuAv6/kMpCAihCvS2AEdebAojxGCzAA3Dep8YCVUjMiExCt364JQp4uxYCO7QTILAl6FC7Dh0zL6Dg8DUxIJQcGNB2ZAzBnUgHYgniVFiFMFAQ8kjAkLjNK4Kqm9bCn1pl6FchjsB7l5BuEOU5DHGKYmYy8RQECIPqP7LaWh2hrmaU/KAEIFYR3/lHGOcUH6Lh4pRYqCck6ezIO0zp48c7sECjzMgxdS5l2qYs8gAB1ewPsGC0PsN0yEwA7b10bgwZuXwQD3LgBc0e5yJowArrA25FMqbCEFlmK+QYlRxXRWXTx84OL/J+EYIE3x6gYoofYAihR2DXkhefO549C65DitkMuaAUkwkPsfLeWidF6IMZYpxYEbFehXu5ExkTIR4Htq4Pq+9VZf08Rwf03B27niYF6WBHcOCD2oP6Ko2QCS8FgRweoNRkGqrKoEXpRhiDpOyIompOSrDku9DfbV3o9T6tIcwC1ggyLelSVAcB7cMlZPtS8sgniAByhCyXtl+AMkwNL1wBqDSGsg2TxhgQwK4eOwzNzTDGSKteG91wmpKQgVlSSD65J+QuRQMaiJz0AUYViNLT5WmgLHSV14u70JEOwc1hRfUdBGfLK8C5B1Wr9lWnqZg/x/wATHKabNoUdKhZY2liL+bIuRLwb0ykOCmE5NwDgQlWTEPzsoOIbMibtAwKgISuAOAcBpfCi+fNgU7tcLwZBJLeD3GUtkTBksMZmmsL/bZ88TC+gXTsmAhS7pQsOpzV03Mt2fppru9GCMwMg1g1BrM+HAGFMBkhpQKHSbvqRZh79IGHV8GgJnL0+g3JbybTHRNMLcBxQYMoLwniIPO0fVmBAomxOifTeJsTkmpOZlk6JgTSh8PCd9LQNT6m1NkA0xprT2nNN6bUws5suHV4N3YL6BjTSlCuXBPBhAYLBNwaMHFWgDmlOQebdwOj4bo370Sdg8gpnG7CeOL6BAtBwuRbtlFqL6aIvxfC3FmLtBPFSxOMlxLGXosJfi0lnLRmNLkDCEfCQmyap1WsRBFQe8kZwOdCeoL7AAOcA4OeGO4hD19WBIasg/ReBwAAkBECgqqvcC8y1bGI4nxaEuA5TqTkXhwnlvMBg1gEM7zil1n6a1N5LAFPpYO0xoweIW1oIp919buQ3MlTtINtv3VrJosIaoyBIHlOKMgpELAURai8JWKs1bEH0MoShABrYFxgGDYGUNAJgxAwcMGIA4Z7UQ3vil8GqdZVg0BRHNTjsiURrknem+QQ6YM+LuM2deAAmhCrjlcTCrHsCuDAD8OLXwZ0g0U/xGqwHcs82cpPSOGLEJT8e14UhwDyuskOx3OdM+u6zniXp/SBk53MtXlg+E8V6L9mb7U3w6QLN1EknAN1k4WmWZaENMq8CW1QQ7hVe68m+iR4EZBzZhrhp4igsstybP+6rUgP0WB2mUPrU28IoA7gwJHvCNIJFb3NjzuA7GN6ghYOCadUtZ0briNEjzS6j0jvY8Ycaa7ForoRWmDMXuCzwzS9LVkVpo9gCpQublERdH6Ktw6AVYkqvjNXmKqZXS2kwr2UfMAJ9IS1mr/cxlZBmV73xLhmlvoz1KAvVYZQIQ70Pq3ENa9M1n2dfq6Xow5fOnnHOEvzBZBCCcA9QSEI2h4q3+gmN7zAWyC+f8436WcBaBavKGHrDgBlJlMuJVVFG+bIYAi+KGbFKtWce5WA+AnmRAv0TdGvfJcFEAjMOKfAhAJAgA2oZGZ9IgkpIgsAmlAbE1MbLAog7DLaSbPeXDBpaVakAiLMM/brZPG/FrNreEMQXgpgHrPrLAx9ESEbC8cbKtXDdbIjDjTbdzRdYwPdfgRjJQfPNyDgDg1wLglgOQnFKReANPUfF0XYbcDyNCQhKwWqW0CwSEcdZnRwolDcCABAAXAsXDfAIIC3YpX0fPUvRQwvYwOKAI56OQqWXw/w4XIIgvNQt3JQ8InQUjUA7/QrMgd5CedvVw8wdw0RTMUwL2ew0RAov5O2MJCJeQd3a8W4BABgeuPkKwYQawagVcSwjiSrC8bwo0HBH/BQ87GACIKANolQ0GTKUYtojQpbOECAaY6Af8WgUSOSJQDgaIpRWpKtLQJYdiS8RpMgVYoVFSMgaCPZewYIWOceQZEwFcGVfSNkZ2IqNcQ46XAqOXceW8PIgCf8dsExHfPeVmGafBZEMQSIkYsYmEX0YgWBYYxY6waBcgXMMUMUbIJAfobIIDUUXIFAfoYUVUCBZg5sYE5aakVwcE0jBErMWEu7C6akpEsgFE1EpUMWFAXUQMbIYUeUAk8gIkgrfot5Esc8SwSRC0AE6wFwjwgo1OIozZfpW0awIid42XLKRPLor0c2GrcglfE4CEsFPwxQCEkpck8ExDBEytUgm3VMA0uIyY4pZBU060ywBEzYo0VgqbQreyDqd8EnV4d4WzGw+wDU9NAbFRWgKXCAR3OXTkYqbAT7L4CsVAGedgGldRTQqzZjAM9wjgZEUQMgUHetfeD2SEAALTgHylXHCR2BgG4D6NanigCnjJij13IHnFPiiVwDXih2lTtnjLqkXigAdGwG2FTkPidUbMSnwGClSjCiIgQGyk5wSkCmnOShCjSiIloEXIVxImPEGVYkgm+Q9j+HBDuP91Xg6Lb3YhzXrM0jm19K9MnJXPXLnIXPuFxVDJVKOx+WKhY1s2UAxCbPwCN0FIuCApfIEi3PfIXAjK/LbHHj/MsAAsNinJAqZHRCRlwHuAAEEzAttV1Oki0TFW4FxLMmMyBELU4pSQy4LjtkLMK0L8Qiw2xZ0G5+oyLrNsz4Ni92AK050R1aA69jgG8tpWL8Kwg9yjBD0/wXo94RKpY2L3c7YwtiSMZlLOA7ZH8OA7YBCSQGCboqgEBUSqhe5TLe5jKzLe5VKEZ1LtKH8NL39et9KmSjKTLzK3KxRTLPKrLdTjhbLNKHKBCUBnLDKTKwrzKIrwqTLrKtp/L7K7KBC7VeADKEACQ0r0qMrMqCQYrmw4qtKdLb9cgQqEAlRSqyryqKqlQcrjdkZaBqrjhJtfAdoShYgEgkgYhUh0gq0dpT0/Du4UhUBlAHBcAwgSiZAml1C5LfLHi4xFLvh0xcBQcsxDLKrVqyr6qmRat/w+q+hBqYhEppUYBXAhJ88jZXQOA+KFrQdXpOBtqGB+q9qDqHRjrXRTqRBzraA4orreBdK3SWDRhPgxBtSyAJL+EpLsktLQaYBBltKfrCq/r8QPS/TZsfTGL4IyAdDAyt4Qzky6o20Og4zjQORJSZkwy/hYKoyZdvzIR0yOKvQARwRU5cy5FipCy/EfgyyKysdrBqzdg6zWynzmy0a7IyB2zZ9IR3JuzkQc0So6pLA4BBzhzRz3caArAN5xzw5rwjzOdTzR1nCihlykoUpQoBIFyYow57ASInYCJjBWJYzPsjFEpowyBp48b6ciKo46iFwEAyaTAKboyrsLyTx5BdgbyBaFJ7zhaGzDa+RZzTbJIPzcbc4aVx4GakLALULw6GLcKvIfIihIVIQPbajvhrwfbOd/aqbQ407Y4eQIAM4FwQy6b9J07iAGKWpmLnbZ1pKR0Ug4oeKyBqdeA6bMbLBRD00hL6QatTd+6whraWAx7ZKq15LQNVgtMRMyAUq1q1qNr8Q8qHLmsirkqXKsqT70qd6Tg967LmskqN6XKorIqH7orprL67Zmtgqj7QqPKv6vK3Lz6/KeCAqr7lI+sP7XLvKf7wGfLkCL6AH4rX7lI9LQHjKLLLKzK/6yDOAeLuA6q95GrmrIhWrygOrKhuqiher7rdqUAhqRqxrGNJql7pqfw5r+A8AlrfQVqt7Kr0Hatn0sG7qHqqH9rApDqXqMAIhqJAo28LqvrWHuA4pu6MBuB+HKHlAhHpajqhJxH2QpHMwD6ZHFqv9cGAbxggaZ1yDZ7/4CILq5GQa56LqyAD7DGnJJtEaNBMYgA==) (Entire encoding/decoding is inlined into one method 😋)

**Comment by PaulusParssinen** [comment]
Oh dear, *adds "Investigate LightClient implementation" to todo list*

**Comment by PaulusParssinen** [comment]
There's small corner-case error in the [McLoughin's quantization trick](https://mmcloughlin.com/posts/geohash-assembly) on which this PR is based.

The statement
> Since `y` is in the range `[1,2]`, the largest power of two less than or equal to `y` is `1.0` and the exponent field will always be `1023`.
is wrong for the upper-bound of the range. Exponent bias is `1023` for `[1, 2)` and exponent for `2.0` flips over to `1024`. This means that `90.0` and `180.0` that are clamped to the range maximum `2`, have their signicand (and subsequently, the geohash integer repesentation) zeroed and dequantization will give equivalent output as `-90.0` and `-180.0` respectively.

**Comment by PaulusParssinen** [comment]
And to add to then confusion, the original latitude limits are "wrong". According EPSG:3857 (also known as EPSG:3785, <sub>also known as EPSG:900913.. etc.</sub>) the latitude limits are `-85.05112878` to `85.05112878`.. not `-90` to `90`. Fun! I'm starting to see why the implementation are all over the place as the GeoHash standard uses different limits than the most common web projection.

**Comment by badrishc** [comment]
Interesting stuff! Glad to see you are making progress here. :)

**Comment by PaulusParssinen** [comment]
Added benchmarks from my laptop which has `AVX512F` by which we guard the PEXT/PDEP path. This made both encoding/decoding of geohashes integers go from `~5.5ns` to `~1.5ns` which was more than I expected. It may be worthwhile to enable more platforms with `static readonly` Zen 2 check by doing [`CpuId(0, 0)`](https://apisof.net/catalog/a5674107bdaeccf5d74ed38a1369e139) and ``CpuId(1, 0)`` and check if `AuthenticAMD && family < 0x19` to guard Pre-Zen 3.

Also wrote little bit more about the PR to its description.

**Comment by PaulusParssinen** [comment]
Added more background to the RP description about the nuances of geohash..

File: Garnet.sln
Comment (yrajas) [inline Garnet.sln:1]: Thanks for taking this up. The numbers look impressive.
Do you know if there are any licensing constraints to use these optimizations?

Comment (mmcloughlin) [reply Garnet.sln:1]: > Do you know if there are any licensing constraints to use these optimizations?

Speaking for the quantization optimization derived from my blog post, I didn't explicitly license the blog post itself, though I'd be happy to make any licensing clarifications if necessary? It's public domain as far as I'm concerned.

The optimization is implemented in my MIT-licensed Golang geohash library https://github.com/mmcloughlin/geohash.

Comment (PaulusParssinen) [reply Garnet.sln:1]: If there are any concerns, I'll add third-part notice to the NOTICE.md for both https://github.com/mmcloughlin/geohash and https://github.com/georust/geohash

**Comment by PaulusParssinen** [comment]
Out of curiosity, what required a force-push? 😅

**Review by yrajas — APPROVED** [review]
